library/alloc/src/collections/btree/map.rs

```diff
@@ -1626,6 +1737,36 @@ impl<K, V> Clone for Values<'_, K, V> {
     }
 }
 
+/// An iterator produced by calling `drain` on BTreeMap.
+#[unstable(feature = "btree_drain", issue = "81074")]
+#[derive(Debug)]
+pub struct Drain<K, V> {
+    inner: IntoIter<K, V>,
+}
+
+#[unstable(feature = "btree_drain", issue = "81074")]
+impl<K, V> Iterator for Drain<K, V> {
+    type Item = (K, V);
+
+    fn next(&mut self) -> Option<(K, V)> {
+        self.inner.next()
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.inner.size_hint()
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<K, V> ExactSizeIterator for Drain<K, V> {
```

Comment:
If it is exact, is it also trusted?

Reply:
I trust it… but you probably mean #37306. If I understand correctly, every [ExactSizeIterator](https://doc.rust-lang.org/core/iter/trait.ExactSizeIterator.html) sourcing from a collection (so not adapting like Skip) should be [TrustedLen](https://doc.rust-lang.org/core/iter/trait.TrustedLen.html). But in BTree, none is, so I probably don't.

Reply:
> but you probably mean #37306.

Yes.

> But in BTree, none is, so I probably don't.

So is there any reason not to trust it, can it be wrong?

Reply:
The `Drain` iteraror is a thin wrapper aroind `IntoIter`, albeit instantiated on a map that may well violate some invariants, but only invariants irrelevant for iteration. So the same reasons for distrust hold as why `IntoIter` is not trusted. I have no idea what those reasons could be, for `IntoIter`, for `Drain`, or for any simple iterator over a collection. I imagine it's legal and might even be beneficial to declare a `DrainFilter` iterator as `ExactSizeIterator`, stating it 's intende to be used when the predicate mostly yields true (which is the opposite of the current reality), and then that iterator shouldn't be `TrustedLen` at all. But that's an entirely different situation from `Drain`.

Reply:
> I imagine it's legal and might even be beneficial to declare a DrainFilter iterator as ExactSizeIterator

From a safety perspective it is. From an API contract perspective it is not.

> When doing so, the implementation of Iterator::size_hint must return the exact size of the iterator.